include/proj/common.hh

```diff
@@ -0,0 +1,256 @@
+/******************************************************************************
+ *
+ * Project:  PROJ
+ * Purpose:  ISO19111:2018 implementation
+ * Author:   Even Rouault <even dot rouault at spatialys dot com>
+ *
+ ******************************************************************************
+ * Copyright (c) 2018, Even Rouault <even dot rouault at spatialys dot com>
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a
+ * copy of this software and associated documentation files (the "Software"),
+ * to deal in the Software without restriction, including without limitation
+ * the rights to use, copy, modify, merge, publish, distribute, sublicense,
+ * and/or sell copies of the Software, and to permit persons to whom the
+ * Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included
+ * in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+ * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+ * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ ****************************************************************************/
+
+#ifndef COMMON_HH_INCLUDED
+#define COMMON_HH_INCLUDED
+
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "exception.hh"
+#include "io.hh"
+#include "metadata.hh"
+#include "util.hh"
+
+NS_PROJ_START
+
+/** osgeo.proj.common namespace
+ *
+ * \brief Common classes.
+ */
+namespace common {
+
+// ---------------------------------------------------------------------------
+
+class Date {
+    // TODO
+};
+
+// ---------------------------------------------------------------------------
+
+class UnitOfMeasure {
+  public:
+    enum class PROJ_DLL Type {
+        /** Unknown unit of measure */
+        UNKNOWN,
+        /** No unit of measure */
+        NONE,
+        /** Angular unit of measure */
+        ANGULAR,
+        /** Linear unit of measure */
+        LINEAR,
+        /** Scale unit of measure */
+        SCALE
+    };
+
+    PROJ_DLL UnitOfMeasure(const std::string &nameIn = std::string(),
```

Comment:
Regarding `PROJ_DLL`. Do you aim to fine-tune the exportable interface by selective exporting versus exporting the whole classes?

Reply:
I tried initially to export whole classes but VC emits a lot of warnings due to the hidden "class Private" used for the pimpl paradigm. Exporting only methods avoid that warning.